src/harness/sas_objects.py

```diff
@@ -0,0 +1,387 @@
+                    {grant_response['cbsdId']: cbsdobject})
+
+    def getCbsdObjectById(self, cbsd_id):
+        return self.cbsd_objects[cbsd_id]
+
+    def heartbeatRequestForAllGrants(self):
+        """
+            This function performs heartbeat request for all grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
+            heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
+            for heartbeat_request in heartbeat_requests:
+                heartbeat_response = self.testcase._sas.Heartbeat(
+                    heartbeat_request, self.ssl_cert, self.ssl_key)['heartbeatResponse'][0]
+                grant_object = cbsd_object_item.grant_objects[heartbeat_response['grantId']]
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
+                                                             '%Y-%m-%dT%H:%M:%SZ')
+                    self.testcase.assertLess(datetime.utcnow(), transmit_expire_time)
+                    self.testcase.assertLessEqual(
+                        (transmit_expire_time - datetime.utcnow()).total_seconds(), 240)
+                    grant_object.setState(Grant.AUTHORIZED)
+                else:
+                    self.mapResponseCodeToGrantState(
+                        heartbeat_response['response']['responseCode'], grant_object)
+
+    def mapResponseCodeToGrantState(self, heartbeat_response_code, grant_object):
+        """
+            This function maps the heartbeat response code to the grant state
+        """
+        if (heartbeat_response_code == 502):
+            grant_object.setState(Grant.OUT_OF_SYNC)
+        elif (heartbeat_response_code == 500):
+            grant_object.setState(Grant.TERMINATED)
+        elif (heartbeat_response_code == 501):
+            grant_object.setState(Grant.SUSPENDED)
+        else:
+            assert(not "Invalid grant response code")
+
+
+    def performHeartbeatAndUpdateGrants(self):
+        """
+            This function construct heartbeat request for all active grants and updates the grants
+        """
+        for cbsd_object_item in self.cbsd_objects.values():
+            heartbeat_responses_with_operation_parameters = {}
+
+            #Heartbeat on all active grants
+            heartbeat_requests = cbsd_object_item.constructHeartbeatRequestForAllActiveGrants()
+            for heartbeat_request in heartbeat_requests:
+                heartbeat_response = self.testcase._sas.Heartbeat(
+                    heartbeat_request, self.ssl_cert, self.ssl_key)['heartbeatResponse'][0]
+                grant_object = cbsd_object_item.grant_objects[heartbeat_response['grantId']]
+                if heartbeat_response['response']['responseCode'] == 0:
+                    transmit_expire_time = datetime.strptime(heartbeat_response['transmitExpireTime'],
```

Comment:
We don't need to do these checks: it's not a protocol test case. So in fact you can delegate all of the logic to `mapResponseCodeToGrantState`.

Reply:
Removed   as per suggestion